x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/ExpressionBuilder.java

```diff
@@ -691,62 +691,67 @@ public List<NamedExpression> visitGrouping(EsqlBaseParser.FieldsContext ctx) {
 
     @Override
     public Object visitInputParam(EsqlBaseParser.InputParamContext ctx) {
-        TypedParamValue param = param(ctx.PARAM());
-        DataType dataType = EsqlDataTypes.fromTypeName(param.type);
-        Source source = source(ctx);
-        if (dataType == null) {
-            throw new ParsingException(source, "Invalid parameter data type [{}]", param.type);
-        }
-        if (param.value == null) {
-            // no conversion is required for null values
-            return new Literal(source, null, dataType);
-        }
-        final DataType sourceType;
-        try {
-            sourceType = DataType.fromJava(param.value);
-        } catch (QlIllegalArgumentException ex) {
-            throw new ParsingException(
-                ex,
-                source,
-                "Unexpected actual parameter type [{}] for type [{}]",
-                param.value.getClass().getName(),
-                param.type
-            );
-        }
-        if (sourceType == dataType) {
-            // no conversion is required if the value is already have correct type
-            return new Literal(source, param.value, dataType);
+        QueryParam param = paramByToken(ctx.PARAM());
+        return visitParam(ctx, param);
+    }
+
+    @Override
+    public Object visitInputNamedOrPositionalParam(EsqlBaseParser.InputNamedOrPositionalParamContext ctx) {
+        QueryParam param = paramByNameOrPosition(ctx.NAMED_OR_POSITIONAL_PARAM());
+        if (param == null) {
+            return Literal.NULL;
         }
-        // otherwise we need to make sure that xcontent-serialized value is converted to the correct type
-        try {
+        return visitParam(ctx, param);
+    }
 
-            if (EsqlDataTypeConverter.canConvert(sourceType, dataType) == false) {
-                throw new ParsingException(
-                    source,
-                    "Cannot cast value [{}] of type [{}] to parameter type [{}]",
-                    param.value,
-                    sourceType,
-                    dataType
-                );
-            }
-            return new Literal(source, EsqlDataTypeConverter.converterFor(sourceType, dataType).convert(param.value), dataType);
-        } catch (QlIllegalArgumentException ex) {
-            throw new ParsingException(ex, source, "Unexpected actual parameter type [{}] for type [{}]", sourceType, param.type);
+    private Object visitParam(EsqlBaseParser.ParamsContext ctx, QueryParam param) {
+        Source source = source(ctx);
+        DataType type = param.type();
+        if (type == null) {
+            throw new ParsingException(source, "Invalid parameter data type [{}]", type);
```

Comment:
When is this exception being thrown?

Reply:
This two lines can be removed, as we should always have type(DataType) set in `RequestXContent` now.